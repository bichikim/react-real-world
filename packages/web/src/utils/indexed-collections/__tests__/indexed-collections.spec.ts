import {createIndexedCollections, createIndexedStructure} from '..'

const dataList = [
  {id: 'i1', title: 'bar', type: 'moon'},
  {id: 'i2', title: 'foo', type: 'sun'},
  {id: 'i3', title: 'jhon', type: 'star'},
]

describe('indexed-collections', function test() {
  describe('createIndexedStructure', function test() {
    it('should return a structure for indexing ', function test() {

      const result = createIndexedStructure(dataList, ['id', 'type'])

      expect(result.indexRecord.id).toMatchSnapshot()
      expect(result.indexRecord.type).toMatchSnapshot()
      expect(result.indexRecord.title).toMatchSnapshot()
    })
  })

  describe('createIndexedCollections', function test() {
    it('should return', function test() {

      const result = createIndexedCollections(
        dataList,
        ['id', 'type'],
      )

      expect([...result]).toEqual(dataList)

      expect(result.size).toBe(3)
      expect(result.get('id', 'i1')).toEqual(dataList[0])
      expect(result.has(dataList[0])).toEqual(true)
      expect(result.has({id: 'i4', title: 'dax', type: 'mars'})).toEqual(false)
      expect([...result.keys('type')]).toEqual(['moon', 'sun', 'star'])
      expect([...result.keys('id')]).toEqual(['i1', 'i2', 'i3'])
      expect([...result.values()]).toEqual(dataList)
      expect([...result.entries('id')]).toEqual([
        ['i1', dataList[0]],
        ['i2', dataList[1]],
        ['i3', dataList[2]],
      ])

      const newItem = {
        id: 'i4',
        title: 'dax',
        type: 'mars',
      }

      result.set(newItem)
      expect(result.has(newItem)).toBe(true)

      result.delete(newItem)
      expect(result.has(newItem)).toBe(false)

      result.set(newItem)
      expect(result.has(newItem)).toBe(true)

      result.deleteByKeyValue('id', 'i4')
      expect(result.has(newItem)).toBe(false)

      const list = []

      result.forEach('id', (item, key) => {
        list.push([key, item])
      })

      expect(list).toEqual([
        ['i1', dataList[0]],
        ['i2', dataList[1]],
        ['i3', dataList[2]],
      ])

      expect([...result.map('id', (item, id) => [id, item])]).toEqual([
        ['i1', dataList[0]],
        ['i2', dataList[1]],
        ['i3', dataList[2]],
      ])
    })
  })
})
