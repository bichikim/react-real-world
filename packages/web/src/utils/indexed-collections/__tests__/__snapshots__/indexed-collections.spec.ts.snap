// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createIndexedStructure should return a structure for indexing  1`] = `
Map {
  0 => Object {
    "id": "i1",
    "title": "bar",
    "type": "moon",
  },
  1 => Object {
    "id": "i2",
    "title": "foo",
    "type": "sun",
  },
  2 => Object {
    "id": "i3",
    "title": "jhon",
    "type": "star",
  },
}
`;

exports[`createIndexedStructure should return a structure for indexing  2`] = `
Map {
  "i1" => Object {
    "id": "i1",
    "title": "bar",
    "type": "moon",
  },
  "i2" => Object {
    "id": "i2",
    "title": "foo",
    "type": "sun",
  },
  "i3" => Object {
    "id": "i3",
    "title": "jhon",
    "type": "star",
  },
}
`;

exports[`createIndexedStructure should return a structure for indexing  3`] = `
Map {
  "moon" => Object {
    "id": "i1",
    "title": "bar",
    "type": "moon",
  },
  "sun" => Object {
    "id": "i2",
    "title": "foo",
    "type": "sun",
  },
  "star" => Object {
    "id": "i3",
    "title": "jhon",
    "type": "star",
  },
}
`;

exports[`createIndexedStructure should return a structure for indexing  4`] = `undefined`;

exports[`indexed-collections createIndexedStructure should return a structure for indexing  1`] = `
Map {
  "i1" => Object {
    "id": "i1",
    "title": "bar",
    "type": "moon",
  },
  "i2" => Object {
    "id": "i2",
    "title": "foo",
    "type": "sun",
  },
  "i3" => Object {
    "id": "i3",
    "title": "jhon",
    "type": "star",
  },
}
`;

exports[`indexed-collections createIndexedStructure should return a structure for indexing  2`] = `
Map {
  "moon" => Object {
    "id": "i1",
    "title": "bar",
    "type": "moon",
  },
  "sun" => Object {
    "id": "i2",
    "title": "foo",
    "type": "sun",
  },
  "star" => Object {
    "id": "i3",
    "title": "jhon",
    "type": "star",
  },
}
`;

exports[`indexed-collections createIndexedStructure should return a structure for indexing  3`] = `undefined`;
